# 7. 오류처리
깨끗한 코드와 오류처리는 확실히 연관성이 있다. 상당수 코드 기반은 전적으로 오류 처리 코드에 좌우된다.  
여기서 좌우된다는 표현은 코드 기반이 오류만 처리한다는 의미가 아니다. 여기저기 흩어진 오류 처리 코드 때문에 실제 코드가 하는 일을 파악하기가 거의 불가능하다는 의미다.  
오류 처리는 중요하지만 오류 처리 코드로 인해 프로그램 논리를 이해하기 어려워진다면 깨끗한 코드라 부르기 어렵다.  
이 장에서는 꺠긋하게 오류를 처리하는 기법과 고려 사항 몇 가지를 소개한다.  

### 목차
+ 오류 코드보다 예외를 사용하라
+ Try-Catch-Finally 문부터 작성하라!
+ 미확인(uncheked) 예외를 사용하라!
+ 예외에 의미를 제공하라
+ 호출자를 고려해 예외 클래스를 정의하라.
+ 정상 흐름을 정의하라
+ null을 반환하지 마라
+ null을 전달하지 마라
## 오류 코드보다 예외를 사용하라!
~~~java
// Bad : 오류 코드 사용 
public class DeviceController {
  ...
  public void sendShutDown() {
    DeviceHandle handle = getHandle(DEV1);
    // 디바이스 상태를 점검한다.
    if (handle != DeviceHandle.INVALID) {
      // 레코드 필드에 디바이스 상태를 저장한다.
      retrieveDeviceRecord(handle);
      // 디바이스가 일시정지 상태가 아니라면 죵료한다.
      if (record.getStatus() != DEVICE_SUSPENDED) {
        pauseDevice(handle);
        clearDeviceWorkQueue(handle);
        closeDevice(handle);
      } else {
        logger.log("Device suspended. Unable to shut down");
      }
    } else {
      logger.log("Invalid handle for: " + DEV1.toString());
    }
  }
  ...
}
~~~
+ 호출자 코드가 복잡해진다.(함수를 호출한 즉시 오류를 확인해야 하기 때문)
+ 오류가 발생하면 예외를 던지는 편이 낫다.(논리가 오류 처리 코드와 뒤섞이지 않으니 깔끔해진다.)
~~~java
// Good : 오류를 발견하면 예외를 던지는 코드
public class DeviceController {
  ...
  public void sendShutDown() {
    try {
      tryToShutDown();
    } catch (DeviceShutDownError e) {
      logger.log(e);
    }
  }
    
  private void tryToShutDown() throws DeviceShutDownError {
    DeviceHandle handle = getHandle(DEV1);
    DeviceRecord record = retrieveDeviceRecord(handle);
    pauseDevice(handle); 
    clearDeviceWorkQueue(handle); 
    closeDevice(handle);
  }
  
  private DeviceHandle getHandle(DeviceID id) {
    ...
    throw new DeviceShutDownError("Invalid handle for: " + id.toString());
    ...
  }
  ...
}
~~~
+ 가독성은 물론, **논리 로직**과 **오류처리 로직**을 분리했기 때문에 각 개념을 독립적으로 살펴보고 이해할 수 있다.  

> < Study >  
> 1. try - catch는 public 접근자에서만 주로 사용, private 접근자에서는 사용하지 않는다.
> 2. private에서는 메소드를 만들고, 예외처리는 public으로 빼서 예외처리 하기

## Try-Catch-Finally 문부터 작성하라!
어떤 면에서 try 블록은 트랜잭션과 비슷하다. try 블록에서 무슨 일이 생기든지 catch 블록은 프로그램 상태를 일관성 있게 유지해야 한다.  
그러므로 예외가 발생할 코드를 짤 때는 try-catch-finally 문으로 시작하는 편이 낫다.  
(그러면 try 블록에서 무슨 일이 생기든지 호출자가 기대하는 상태를 정의하기 쉬워진다.)  

예제: 파일을 열어 직렬화된 객체 몇개를 읽어 들이는 코드
~~~java
// 파일이 없으면 예외를 던지는지 알아보는 단위 테스트
// 예외를 던지지 않으므로 단위 테스트는 실패
@Test(expected = StorageException.class)
public void retrieveSectionShoudThrowOnInvalidFileName() {
   sectionStore.retrieveSection("invalid - file");
}
  
public List<RecordedGrip> retrieveSection(String sectionName) {
	// 실제로 구현할 때까지 비어있는 더미를 반환한다.
	return new ArrayList<RecordedGrip>();
}
~~~
~~~java
// 예외를 던지므로 테스트를 통과한다.
public List<RecordedGrip> retrieveSection(String sectionName) {
	try {
      FileInputStream stream = new FileInputStream(sectionName)
    } catch (Exception e) { // < Study > - 너무 광범위하게 예외처리를 하지 않는다
      throw new StorageException("retrieval error", e);
    }
	return new ArrayList<RecordedGrip>();
}
~~~
~~~java
// 리팩터링
// catch 블록에서 예외 유형을 좁혀, 실제로 FileInputStream 생성자가 던지는 FileNotFoundException을 잡아낸다.
public List<RecordedGrip> retrieveSection(String sectionName) {
	try {
	  FileInputStream stream = new FileInputStream(sectionName);
	  stream.close();
	} catch (FileNotFoundException e) {
	  throw new StorageException("retrieval error", e);
	}
	return new ArrayList<RecordedGrip>();
}
 ~~~
 > < Study >  
> + Scope 범위(동작의 범위) : { ~~~ }
>	+ 트랜잭션으로 보면 시작부터 끝까지라 할 수 있음
> + return new ArrayList<~>();처럼 `빈값을 리턴할 떄는 0`을 넣어 주는게 좋음
> + public 메소드는 남발하지 말 것
>  + 자기방어를 꼭 해야함  
> + 예외에 대한 처리(try / catch), 로그 찍는 코드 등은 public에서 처리한다.
>  	+ private에서는 처리 x, throws로 public에 넘긴다.

## 미확인(uncheked) 예외를 사용하라!
###### Checked exception VS Unchecked Exception
+ Checked exception가 치르는 비용에 상응하는 이익을 제공하는지 따져봐야한다.
	+ Checked exception은 OCP(Open Closed principle, 개방 폐쇄 원칙)을 위반하다.
		1. 특정 메소드에서 checked exception을 throw하고
		2. 3단계(메소드 콜) 위의 메소드에서 그 exception을 catch한다면
		3. 모든 중간단계 메소드에 exception을 정의해야 한다.(자바의 경우 메소드 선언에 throws 구문을 붙이는 등)
			+ 상위 레벨 메소드에서 하위 레벨 메소드의 디테일(예외 해)에 대해 알아야 하기 때문에 캡슐화 또한 깨진다.

필요한 경우 checked exceptions를 사용해야 되지만 일반적인 경우 득보다 실이 많다.  

[Checked Exception & Unchecked Exception이란?](https://close852.tistory.com/47)

## 예외에 의미를 제공하라
예외를 던질 떄는 전후 상황을 충분히 덧붙인다. 그러면 오류가 발생한 원인과 위치를 찾기가 쉬워진다.  
즉, 오류 메시지에 정보를 담아 예외와 함께 던진다.  
+ 실패한 연산 이름과 실패 유형도 언급  

애플리케이션이 로깅 기능을 사용한다면 catch 블록에서 오류를 기록하도록 충분한 정보를 넘겨준다.

## 호출자를 고려해 예외 클래스를 정의하라.
+ 예외 클래스에서 제공하는 정보의 초점은 `오류를 잡아내는 방법`이 되어야 한다.
+ 서드파티 라이브러리를 사용하는 경우 그것들을 wrapping함으로써 라이브러리 교체 등의 변경이 있는 경우 대응하기 쉬워진다.  
~~~java
// Bad
// 코드의 중복이 심함
  
  ACMEPort port = new ACMEPort(12);
  try {
    port.open();
  } catch (DeviceResponseException e) {
    reportPortError(e);
    logger.log("Device response exception", e);
  } catch (ATM1212UnlockedException e) {
    reportPortError(e);
    logger.log("Unlock exception", e);
  } catch (GMXError e) {
    reportPortError(e);
    logger.log("Device response exception");
  } finally {
    ...
  }
	/*
    		< Study >
    		위의 코드처럼, 예외에 따른 로직이 동일할 경우, 각각의 예외로 처리하지 말고
    		아래의 코드처럼 상위 예외 클래스를 만들어서 묶어주자. 
    	 */
~~~
~~~java
// Good
// LocalPort 클래스는 ACMEPort 클래스가 던지는 예외를 잡아 변환하는 wrapper 클래스
  LocalPort port = new LocalPort(12);
  try {
    port.open();
  } catch (PortDeviceFailure e) {
    reportError(e);
    logger.log(e.getMessage(), e);
  } finally {
    ...
  }
  
  public class LocalPort {
    private ACMEPort innerPort;
    public LocalPort(int portNumber) {
      innerPort = new ACMEPort(portNumber);
    }
    
    public void open() {
      try {
        innerPort.open();
      } catch (DeviceResponseException e) {
        throw new PortDeviceFailure(e);
      } catch (ATM1212UnlockedException e) {
        throw new PortDeviceFailure(e);
      } catch (GMXError e) {
        throw new PortDeviceFailure(e);
      }
    }
    ...
  }
~~~

## 정상 흐름을 정의하라
앞 절에서 충고한 지침을 따른다면 비즈니스 논리와 오류 처리가 잘 분리된 코드가 나온다.  
하지만 그러다 보면 오류 감지가 프로그램 언저리로 밀려난다. 외부 API를 감싸 독자적인 예외를 던지고, 코드 위에 처리기를 정의해 중단된 계산을 처리한다.  
대개는 멋진 처리 방식이지만, 떄로는 중단이 적합하지 않을 때도 있다.
~~~java
// Bad
 try {
    MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
    m_total += expenses.getTotal();
  } catch(MealExpensesNotFound e) {
    m_total += getMealPerDiem();
  }
	/*
      < Study >
      위의 코드처럼 null일 경우 단순한 예외로만 처리할 경우 그에 따른 추가 코드가 필요하다.
      따라서 밑의 코드처럼 null일 경우 발생하는 에러에 따른 default 값을 만들어 주자.
    */
~~~
위에서 식비를 비용으로 청구했다면 직원이 청구한 식비를 총계에 더한다. 식비를 비용으로 청구하지 않았다면 일일 기본 식비를 총계에 더한다.  
그런데 예외가 논리를 따라가기 어렵게 만든다. 따라서 특수 상황을 처리할 필요가 없다면 더 좋을 것이다.(코드가 훨씬 더 간결해짐)  
~~~java
 // Good

  // caller logic.
  ...
  MealExpenses expenses = expenseReportDAO.getMeals(employee.getID());
  m_total += expenses.getTotal();
  ...
  
  public class PerDiemMealExpenses implements MealExpenses {
    public int getTotal() {
      // return the per diem default
    }
  }
  
  // 이해를 돕기 위해 직접 추가한 클래스
  // https://github.com/Yooii-Studios
  public class ExpenseReportDAO {
    ...
    public MealExpenses getMeals(int employeeId) {
      MealExpenses expenses;
      try {
        expenses = expenseReportDAO.getMeals(employee.getID());
      } catch(MealExpensesNotFound e) {
        expenses = new PerDiemMealExpenses();
      }
      
      return expenses;
    }
    ...
  }
~~~
클래스를 만들거나 객체를 조작해 특수 사례를 처리하는 방식이다. 이를 특수 사례 패턴(Special Case Pattern)이라 부른다.  
클래스나 객체가 예죅인 상황을 캡슐화해서 처리하므로, 클라이언트 코드가 예외적인 상황을 처리할 필요가 없어진다.

## null을 반환하지 마라
우리가 흔히 저지르는 바람에 오류를 유발하는 행위
+ null을 반환하는 습관
	~~~java
	/*
		Bad
		한 줄 건너 하나씩 null을 확인하는 코드
	*/
	public void registerItem(Item item) {
		if (item != null) {
			ItemRegistry registry = persistentStore.getItemRegistry();
			if (registry != null) {
				....
			}
		}
	}
	~~~
	null을 반환하는 코드는 일거리를 늘릴 뿐만 아니라 호출자에게 문제를 떠넘긴다.   
	누구 하나라도 null 확인을 빼먹는다면 애플리케이션이 통제 불능에 빠질지도 모른다.  
	메서드에서 null을 반환하고픈 유혹이 든다면 그 대신에 `예외를 던지거나 특수 사례 객체를 반환한다.`  
	(사용하려는 외부 API가null을 반환한다면 감싸기 메서드를 구현해 예외를 던지거나 특수 사례 객체를 반환하는 방식을 고려한다.)  
	~~~java
	 // Bad
      List<Employee> employees = getEmployees();
      if (employees != null) {
        for(Employee e : employees) {
          totalPay += e.getPay();
        }
      }
     /*
   	 	< Study >
  	 	Collection일 떄는 null을 리턴하지 말자!
 	 	- isEmpty()가 있으니, 활용할 수 있는 발판을 만들자
     */
	~~~
	~~~java
	  /*
  	  	Good
 	  	getEmployees를 변경해 빈 리스트를 반환한다면 코드가 훨씬 깔끔해질뿐더러 NullPointerException이 발생할 가능성도 줄어든다.
 	   */
      List<Employee> employees = getEmployees();
      for(Employee e : employees) {
        totalPay += e.getPay();
      }
      
      public List<Employee> getEmployees() {
        if( .. there are no employees .. )
          return Collections.emptyList();
        }
	~~~
	
## null을 전달하지 마라
메서드에서 null을 반환하는 방식도 나쁘지만, 메서드로 null을 전달하는 방식은 더 나쁘다.  
정상적인 인수로 null을 기대하는 API가 아니라면, 메서드로 null을 전하는 코드는 최대한 피한다.  
~~~java
// Bad
// calculator.xProjection(null, new Point(12, 13));
// 위와 같이 부를 경우 NullPointerException 발생
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    return (p2.x – p1.x) * 1.5;
  }
  ...
}

// Bad
// NullPointerException은 안나지만 위 단계에서 InvalidArgumentException이 발생할 경우 처리해줘야 함.
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    if(p1 == null || p2 == null){
      throw InvalidArgumentException("Invalid argument for MetricsCalculator.xProjection");
    }
    return (p2.x – p1.x) * 1.5;
  }
}

// Bad
// 좋은 명세이지만 첫번째 예시와 같이 NullPointerException 문제를 해결하지 못한다.
public class MetricsCalculator {
  public double xProjection(Point p1, Point p2) {
    assert p1 != null : "p1 should not be null";
    assert p2 != null : "p2 should not be null";
    
    return (p2.x – p1.x) * 1.5;
  }
}
~~~
가장 좋은 방법은 애초에 null을 넘기지 못하도록 금지하자

> < Study >  
> + null 값 체크는 private에서 하지 말자!
> 	+ private 메소드를 호출하는 public 메소드에서 null 체크가 선행되어야 한다.

[Assert란?](https://runtoyourdream.tistory.com/223)